src/compiler/checker.ts

```diff
@@ -23411,7 +23411,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     const sourceArity = getTypeReferenceArity(source);
                     const targetArity = getTypeReferenceArity(target);
                     const sourceRestFlag = isTupleType(source) ? source.target.combinedFlags & ElementFlags.Rest : ElementFlags.Rest;
-                    const targetRestFlag = target.target.combinedFlags & ElementFlags.Rest;
+                    const targetHasRestElement = target.target.hasRestElement;

```

Comment:
`type.hasRestElement = !!(combinedFlags & ElementFlags.Variable);`, and `ElementFlags.Variable` is `ElementFlags.Rest | ElementFlags.Variadic` so... yeah, I guess this works to expand the check from just rest to rest and variadic elements! I read this like 3 times before checking how `hasRestElement` was defined and realizing that, though.

We should slap a `@deprecated` on `hasRestElement`, really.  The element flags are really the new way to check that.

Reply:
> We should slap a @deprecated on hasRestElement, really. The element flags are really the new way to check that.

https://github.com/microsoft/TypeScript/pull/58864 😉